santoku/utils/configuration.py

```diff
@@ -0,0 +1,347 @@
+    Here's how the schema of the above example would have to look like:
+
+    {
+        "name": {"type": "string"},
+        "settings": {
+            "type": "object",
+            "properties": {
+                "boolean_setting": {"type": "boolean"},
+                "int_setting": {"type": "integer"},
+                "object_setting": {
+                    "type": "object",
+                    "properties": {
+                        "nested_float_setting": {"type": "number"},
+                        "nested_string_setting": {"type": "string"}
+                    }
+                }
+            }
+        }
+    }
+
+    Configurations must follow the schema if given.
+    """
+
+    JSON = Union[dict, list]  # naive type hint for parsed JSON objects
+
+    def __init__(
+        self,
+        configurations: List[dict] = None,
+        schema: JSON = None,
+        initial_configuration: str = None,
+    ):
+        """
+        A list of available `configurations` and a `schema` can be passed to be stored in the
+        manager. The `initial_configuration` is an optional value. It must be the name of one of
+        the `configurations`, which will become the active configuration. This is simply to save
+        the user the initial call to `apply_configuration`.
+
+        Parameters
+        ----------
+        configurations : List[Dict[str, Union[str, Dict[str, Any]]]], optional
+            Collection of configurations with different combination of settings.
+        schema : Dict[str, Any], optional
+            Structure of the settings with the corresponding types of each setting value.
+        initial_configuration: str, optional
+            Name of the configuration in the `configurations` that will be set as the initial one.
+
+        InvalidConfiguration
+            If one or more elements in `configurations` does not contain the "name" or "settings" keys.
+
+        Notes
+        -----
+        The schema must conform to draft 7 of the JSON Schema specification [1].
+        The following structure is assumed for the list of predefined configurations:
+            [
+                {
+                    "name": <name_of_the_configuration>,
+                    "settings": {
+                        <key>: <value>,
+                        ...
+                    }
+                },
+                ...
+            ]
+
+        See also
+        --------
+        validate_schema
+            Called to check whether each configuration follow the schema (if provided).
+        define_configuration
+            Called to add each configuration to the list of available configurations.
+        apply_configuration
+            Called to activate the `initial_configuration` (if provided).
+
+        References
+        ----------
+        [1] :
+        https://json-schema.org/specification-links.html#draft-7
+        """
+        if schema:
+            Draft7Validator.check_schema(schema=schema)
```

Comment:
I guess this is throwing an exception. We leave that one to bubble?

Reply:
Yes, all validators do nothing if validation is successful and throw exceptions if the validation fails. All in all it makes sense to let their more useful error messages bubble than to capture them.